gcc/rust/analysis/rust-hir-liveness.cc

```diff
@@ -0,0 +1,154 @@
+// the terms of the GNU General Public License as published by the Free
+// Software Foundation; either version 3, or (at your option) any later
+// version.
+
+// GCC is distributed in the hope that it will be useful, but WITHOUT ANY
+// WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+// for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with GCC; see the file COPYING3.  If not see
+// <http://www.gnu.org/licenses/>.
+
+#include "rust-hir-liveness.h"
+#include "rust-hir-full.h"
+#include "rust-name-resolver.h"
+
+extern bool
+saw_errors (void);
+
+namespace Rust {
+namespace Analysis {
+
+class FindEntryPoint : public LivenessBase
+{
+  using Rust::Analysis::LivenessBase::visit;
+
+public:
+  static std::vector<HirId> find (HIR::Crate &crate)
+  {
+    FindEntryPoint findEntryPoint;
+    for (auto it = crate.items.begin (); it != crate.items.end (); it++)
+      {
+	it->get ()->accept_vis (findEntryPoint);
+      }
+    return findEntryPoint.getEntryPoint ();
+  }
+
+  // TODO not only fn main can be a entry point.
+  void visit (HIR::Function &function) override
+  {
+    if (function.get_function_name () == "main")
+      {
+	entryPoints.push_back (function.get_mappings ().get_hirid ());
+      }
+  }
+
+private:
+  FindEntryPoint () : LivenessBase () {}
+  std::vector<HirId> entryPoints;
+  std::vector<HirId> getEntryPoint () { return entryPoints; }
+};
+
+std::set<HirId>
+Liveness::Analysis (HIR::Crate &crate)
+{
+  Liveness liveness (FindEntryPoint::find (crate));
+  liveness.go (crate);
+
+  return liveness.liveSymbols;
+}
+
+void
+Liveness::go (HIR::Crate &crate)
+{
+  while (!worklist.empty ())
+    {
+      HirId hirId = worklist.back ();
+      worklist.pop_back ();
+      scannedSymbols.emplace (hirId);
+      HIR::Item *item
+	= mappings->lookup_hir_item (crate.get_mappings ().get_crate_num (),
+				     hirId);
+      if (item == nullptr)
+	{
+	  std::cout << "nullptr" << std::endl;
+	  continue;
+	}
+      liveSymbols.emplace (hirId);
+      item->accept_vis (*this);
+    }
+  // for (auto it = liveSymbols.begin(); it != liveSymbols.end(); it++) {
+  //     HirId hirId = *it;
+  //     HIR::Item *item = mappings->lookup_hir_item
+  //     (crate.get_mappings().get_crate_num(), hirId);
+  //     // std::cout << item->get_mappings().as_string() << std::endl;
+  //     std::cout << *it << std::endl;
+  // }
+}
+
+void
+Liveness::visit (HIR::ExprStmtWithoutBlock &stmt)
+{
+  stmt.get_expr ()->accept_vis (*this);
+}
+
+void
+Liveness::visit (HIR::CallExpr &expr)
+{
+  expr.get_fnexpr ()->accept_vis (*this);
+}
+
+void
+Liveness::visit (HIR::PathInExpression &expr)
+{
+  NodeId ast_node_id = expr.get_mappings ().get_nodeid ();
+  NodeId ref_node_id = UNKNOWN_NODEID;
+  if (resolver->lookup_resolved_name (ast_node_id, &ref_node_id))
```

Comment:
This piece is a bit confusing with the definition piece, I am currently working on cleaning up how this works well done for getting this bit right.